src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+			req_len = sizeof ( nbd->rx_neg_init );
+			break;
+
+		case NBD_RX_NEG_EXP_NAME:
+			rx = nbd_rx_neg_exp_name;
+			req_len = sizeof ( nbd->rx_exp_name_reply );
+			break;
+
+		case NBD_RX_NEG_OPT_INFO:
+			rx = nbd_rx_neg_opt_info;
+			req_len = sizeof ( nbd->rx_opt_reply );
+			break;
+
+		case NBD_RX_NEG_REP_INFO:
+			rx = nbd_rx_neg_rep_info;
+			req_len = nbd->reply_info_length;
+			break;
+
+		case NBD_RX_TRANS_REP_CMD:
+			rx = nbd_rx_trans_rep_cmd;
+			req_len = sizeof ( nbd->rx_trans_reply );
+			break;
+
+		case NBD_RX_TRANS_DATA:
+			rc = nbd_rx_trans_data ( nbd, iobuf );
+			goto done;
+
+		default:
+			assert ( 0 );
+			rc = -EINVAL;
+			goto done;
+		}
+
+		DBGCIO ( nbd, "NBD %p state %d req %zd, off %zd, \n",
+			 nbd, nbd->rx_state, req_len, nbd->rx_offset );
+
+		part_len = req_len - nbd->rx_offset;
+		if ( part_len > iob_len ( iobuf ) )
+			part_len = iob_len ( iobuf );
+		memcpy ( nbd->rx_buffer + nbd->rx_offset,
+			 iobuf->data, part_len );
+		nbd->rx_offset += part_len;
+
+		iob_pull ( iobuf, part_len );
+
+		/* If all the data for this state has not yet been
+		 * received, stay in this state for now.
+		 */
+		if ( nbd->rx_offset != req_len ) {
+			rc = 0;
+			goto done;
+		}
+
+		rc = rx ( nbd );
+		if ( rc ) {
+			DBGC ( nbd, "NBD %p could not process received "
+			       "data: %s\n", nbd, strerror ( rc ) );
+			goto done;
+		}
+
+		nbd->rx_offset = 0;
+	}
+
+ done:
+	/* Free I/O buffer */
+	free_iob ( iobuf );
+
+	/* Destroy session on error */
+	if ( rc != 0 )
+		nbd_close ( nbd, rc );
+
+	return rc;
+}
+
+/**
+ * Handle data transfer window change
+ *
+ * @v nbd		NBD session
+ */
+static void nbd_socket_window_changed ( struct nbd_session *nbd ) {
+	DBGCP ( nbd, "NBD %p socket window changed\n", nbd );
+	nbd_tx_resume ( nbd );
+}
+
+/** NBD socket interface operations */
+static struct interface_operation nbd_socket_op[] = {
+	INTF_OP ( xfer_deliver, struct nbd_session *, nbd_socket_deliver ),
+	INTF_OP ( xfer_window_changed, struct nbd_session *,
+		  nbd_socket_window_changed ),
+	INTF_OP ( intf_close, struct nbd_session *, nbd_close ),
+};
+
+/** NBD socket interface descriptor */
+static struct interface_descriptor nbd_socket_desc =
+	INTF_DESC ( struct nbd_session, socket, nbd_socket_op );
+
+
+/**
+ * Open NBD URI
+ *
+ * @v parent		Parent interface
+ * @v uri		URI
+ * @ret rc		Return status code
+ */
+static int nbd_open ( struct interface *parent, struct uri *uri ) {
+	struct nbd_session *nbd;
+	char *name;
+	struct sockaddr_tcpip server;
+	int rc;
+
+	/* Sanity check */
+	if ( ! uri->host )
+		return -EINVAL;
+	if ( ! uri->path )
+		return -EINVAL;
+
+	/* Allocate and initialise structure */
+	nbd = zalloc ( sizeof ( *nbd ) );
+	if ( ! nbd )
+		return -ENOMEM;
+
+	ref_init ( &nbd->refcnt, nbd_free );
+	intf_init ( &nbd->block, &nbd_block_desc, &nbd->refcnt );
+	intf_init ( &nbd->socket, &nbd_socket_desc, &nbd->refcnt );
+
+	nbd->uri = uri_get ( uri );
+	name = strchr ( uri->path, '/' );
```

Comment:
For a URI with a non-NULL host and path, `uri->path` should always start with a '/'.  What circumstances did you encounter in which this was not the case?

Reply:
I modified the code so that if the path is not set, an empty name will be used (this is allowed by the NBD protocol), otherwise the leading slash will be omitted.